lib/ecto/query/planner.ex

```diff
@@ -1264,40 +1273,69 @@ defmodule Ecto.Query.Planner do
+  defp prewalk({:selected_as, [], [name]}, _kind, query, expr, counter, _adapter) do
     name = selected_as!(query.select.aliases, name)
-    {{:selected_as, [], [name]}, acc}
+    {{:selected_as, [], [name]}, expr, counter}
+  end
+
+  defp prewalk({:map, _, [{:&, _, [ix]} = amp]}, kind, query, expr, counter, _adapter) do
+    fields = map_fields!(kind, query, ix)
+    new_take = Select.add_take(expr.take, ix, {:map, fields})
+    {amp, %{expr | take: new_take}, counter}
   end
 
-  defp prewalk(%Ecto.Query.Tagged{value: v, type: type} = tagged, kind, query, expr, acc, adapter) do
+  defp prewalk(%Ecto.Query.Tagged{value: v, type: type} = tagged, kind, query, expr, counter, adapter) do
     if Ecto.Type.base?(type) do
-      {tagged, acc}
+      {tagged, expr, counter}
     else
-      {dump_param(kind, query, expr, v, type, adapter), acc}
+      {dump_param(kind, query, expr, v, type, adapter), expr, counter}
     end
   end
 
-  defp prewalk({left, right}, kind, query, expr, acc, adapter) do
-    {left, acc} = prewalk(left, kind, query, expr, acc, adapter)
-    {right, acc} = prewalk(right, kind, query, expr, acc, adapter)
-    {{left, right}, acc}
+  defp prewalk({left, right}, kind, query, expr, counter, adapter) do
+    {left, expr, counter} = prewalk(left, kind, query, expr, counter, adapter)
+    {right, expr, counter} = prewalk(right, kind, query, expr, counter, adapter)
+    {{left, right}, expr, counter}
+  end
+
+  defp prewalk({left, meta, args}, kind, query, expr, counter, adapter) do
+    {left, expr, counter} = prewalk(left, kind, query, expr, counter, adapter)
+    {args, expr, counter} = prewalk(args, kind, query, expr, counter, adapter)
+    {{left, meta, args}, expr, counter}
+  end
+
+  defp prewalk(list, kind, query, expr, counter, adapter) when is_list(list) do
+    {value, {expr, counter}} =
+      Enum.map_reduce(list, {expr, counter}, fn value, {expr, counter} ->
+        {value, expr, counter} = prewalk(value, kind, query, expr, counter, adapter)
+        {value, {expr, counter}}
+      end)
+
+    {value, expr, counter}
   end
 
-  defp prewalk({left, meta, args}, kind, query, expr, acc, adapter) do
-    {left, acc} = prewalk(left, kind, query, expr, acc, adapter)
-    {args, acc} = prewalk(args, kind, query, expr, acc, adapter)
-    {{left, meta, args}, acc}
+  defp prewalk({{:., dot_meta, [left, field]}, meta, []},  kind, query, expr, counter, _adapter) do
+    {ix, ix_expr, ix_query} = get_ix!(left, kind, query)
+    extra = if kind == :select, do: [type: type!(kind, ix_query, expr, ix, field)], else: []
+    field = field_source(get_source!(kind, ix_query, ix), field)
+    {{{:., extra ++ dot_meta, [ix_expr, field]}, meta, []}, expr, counter}
   end
 
-  defp prewalk(list, kind, query, expr, acc, adapter) when is_list(list) do
-    Enum.map_reduce(list, acc, &prewalk(&1, kind, query, expr, &2, adapter))
+  defp prewalk(other, _kind, _query, expr, counter, _adapter) do
+    {other, expr, counter}
```

Comment:
Why did we need those changes to prewalker? In any case, the `acc` is opaque, so if you need a counter, you should include the counter in the accumulator. :)

Reply:
I think I understand. `counter` is the previous `acc` but now you need to return the `take`. Let me think about it.

Reply:
Here is any idea: what if instead of changing `prewalk` to expand `map/1`, you change `collect_fields` to consider `map/1`? You can test it out this way:

1. Remove the `{:map, _, [_]}` clause from prewalk
2. Add the clause to `collect_fields`
3. If it works, you can revert all changes to prewalk (probably easier to revert the file :D)

Reply:
oh actually I tried the collect_fields route first and then I noticed something broke here https://github.com/elixir-ecto/ecto/blob/8c232ccc7169f1c18bd74966d9993e95f67b7e63/lib/ecto/query/planner.ex#L1387 particularly with preloads. i'll reproduce to get your opinion on it

Reply:
Hrm, it seems I should be the one giving it a try then. :)

Reply:
Sorry I couldn't remember the exact issue. I think it `was that when you do something like this

`from p in Post, join: c in Comment, select: map(p), preload: [:comments]`

the `from` take doesn't know it's a map so it errors trying to find `__struct__` somewhere in the post processor. But for `map/2` it knows it's a map because it's tagged that way in the `take` struct.

I could reproduce later tonight (my time) to make it easier for you. During the day I got a baby trying to press my computer key though =) so can't do anything too complicated atm.

Reply:
I think there was also an issue with missing fields because the `from` take didn't know about them.

Reply:
@greg-rychlewski perfect, no rush. if you can push a failing test case (which should be working right now), then it would help me explore this. :)

Reply:
I created a new branch in my fork with the following:

1. Don't expand map/1 while calling the select merge function. Wait until the planner.
2. Don't expand map/1 while prewalking but handle it during `collect_fields`
3. Added a couple integration tests that pass in this branch but not with these changes

Please see here: https://github.com/greg-rychlewski/ecto/commit/0b40a751bcef7a143294059c77afd428a301746f

I trie to make it as clear as possible with my comments but please let me know if I can explain better. Thank you for looking at it.

Reply:
Thank you so much. We should definitely include those tests in the final merge!

I think the root cause of all or our problems is here:

https://github.com/elixir-ecto/ecto/pull/4139/files#diff-a4782321bba4a264f78ae4d81bc9be152d20af799129133dd078d59f2c1fb4d0R117-R118

Note that the handling of `map/struct` fully happens on the escape function. However, because we don't have access to fields here, we keep `map` as is and that's what's pushing complexity everywhere.

Maybe we can try this implementation?

```
  defp escape({tag, _, [{var, _, context}]}, {params, acc}, vars, env)
       when tag in [:map, :struct] and is_atom(var) and is_atom(context) do
    expr = Builder.escape_var!(var, vars)
    take = add_take(acc.take, Builder.find_var!(var, vars), {tag, :all})
    {expr, {params, %{acc | take: take}}}
  end
```

The idea is that we will have this `:all` annotation, which we will have to handle downstream, but perhaps without needing to change `prewalk`. We will no longer need `expand_map` but we will need to be able to merge `:all` (merging `anything` with all returns `anything`).

WDYT?

This is one of the most complex parts of Ecto, which makes doing a small change really hard, but it also one of the reasons why making a small change is important!

Reply:
I am also thinking we should remove the `take` system altogether. Instead of doing:

    %{:&, [], [ix]}

We could keep it in the AST as:

    {:map, _, [%{:&, [], [ix]}, [:foo, :bar]]}

In the planner, every time we see `%{:&, [], [ix]}` we immediately fetch its taken value anyway, and now we will no longer need to because it is immediately available. In `Select.merge`, we already have to consider both the expression and its take anyway, and we could then do it in a single pass.

Perhaps the smarter thing to do is to find a new representation altogether. We have four possible values:

  * sources
  * structs
  * maps
  * merge

And maybe we could represent them as (at the AST level)

  * source(ix, :map | :struct | :any, fields | :all)
  * struct(name, fields)
  * map(fields)
  * merge(any_of_the_four, map)

This should reduce the amount of back and forth between the merger, planner, and the postprocessor and the number of distinct representations.

Reply:
If we don't want to rock the boat too much we can keep the representation as is today except that:

* `&ix` - represents :any (of map or struct)
* `map(&ix, fields)` - represents map
* `struct(&ix, fields)` - represents struct
* `%{...}` - map
* `%name{...}` - struct
* `merge(left, right)`

It is the same as above except more flattened. If we move to this structure, then we can say `map/1` is equivalent to `map(&ix, :all)`.

Reply:
Sorry need to parse the other messages but I could reply to this one now :)

> The idea is that we will have this :all annotation, which we will have to handle downstream, but perhaps without needing to change prewalk. We will no longer need expand_map but we will need to be able to merge :all (merging anything with all returns anything).
> WDYT?

Oh yes actually this is similar to what I had first except I used `nil` instead of `:all` because sometimes `Access.fetch` is called on it. So that makes it work in those places without an extra case.

If you're ok with this being a possibility I could for sure bring that version back to look at. It's been a while but over here I think you weren't too thrilled about it https://github.com/elixir-ecto/ecto/pull/4139#issuecomment-1487571422

But if it is looking better now I could bring it back for consideration.

Reply:
>This should reduce the amount of back and forth between the merger, planner, and the postprocessor and the number of distinct representations.

Reply:
Awesome. So let's give a try at the refactoring. I was planning to give it a go but, if you could do it, even better, cause then we reduce the bus number. According to classify_merge, we have three cases:

  * &ix
  * %{...} - map
  * %name{...} - struct
  * merge(left, right)

BUT the first case actually breaks in three:

  * &ix - no map or struct specified in take
  * &ix - but map specified in take
  * &ix - but struct specified in take

My suggestion is to, when we see map/2 or struct/2 in escape (the last two cases above), we will keep keep the `map/2` and `struct/2` in the AST. Once we start doing this, you will need to change both merge!, subquery, and collect_fields, but once you do so, then we should no longer need `take`. There are some cleanups I see we can do before hand, so I will give them a try first.

Reply:
@greg-rychlewski I pushed two small fixes and some code comments to collect fields. I made `merge` itself recursive and removed an outdated `:never` check. This means that supporting `map/2` and `struct/2` on collect fields is easier, as we only need to add two clauses (one for map and another for struct) instead of several closes that consider how `merge` and other factors interplay.

Reply:
beautiful changes! I'l give the refactoring a try and push them here

Reply:
Let’s do the refactor in a separate PR and then add this functionality later so we don’t mix them. :)

Reply:
Btw sorry I forgot to ask yesterday. Do you mean completely remove the take system or just for map/structs? For instance for this:

```elixir
from p in Post, select: [:title, :id]
```

Would we want to represent it as something like this `{:source, _, [{:&, _, [ix]}, [:title, :id]]}`? I like the idea of completely getting rid of take but not sure if you saw a reason to keep this one.

Reply:
We will keep the system but instead of storing it as a separate field we keep all info in the AST. We can rewrite the case you posted as `map_or_struct(&ix, […])` (if we have :all we can wrap all sources in that but maybe once we do it we will see the benefits of keeping it separate).